R/join_keys.R

```diff
@@ -0,0 +1,651 @@
+#' create a new object.
+#'
+#' Note that join keys are created symmetrically, that is, if `dat1` and `dat2`
+#' have a join key of `col1`, then 2 join keys are created, `dat1 → dat2` and
+#' `dat2 → dat1`. The only exception is for a primary key.
+#'
+#' @param ... (optional), when no argument is given the empty constructor is called.
+#' Otherwise, when called with only one argument of type: `join_keys` or  `teal_data`
+#' it will return the `join_keys` of that object.
+#' When called with 1 or more `join_key_set` it will create a new `join_keys` with
+#' constructed from the arguments.
+#'
+#' @return `join_keys` object.
+#'
+#' @export
+#'
+#' @examples
+#' # Setting join keys ----
+#'
+#' jk <- join_keys(
+#'   join_key("dataset_A", "dataset_B", c("col_1" = "col_a")),
+#'   join_key("dataset_A", "dataset_C", c("col_2" = "col_x", "col_3" = "col_y"))
+#' )
+#' jk
+#'
+#' # or
+#' jk <- join_keys()
+#' jk[["dataset_A"]][["dataset_B"]] <- c("col_1" = "col_a")
+#' jk[["dataset_A"]][["dataset_C"]] <- c("col_2" = "col_x", "col_3" = "col_y")
+#' jk
+#'
+#' td <- teal_data(join_keys = join_keys(join_key("a", "b", "c")))
+#' join_keys(td)
+#'
+#' jk <- join_keys()
+#' join_keys(jk)
+#'
+#' jk <- join_keys()
+#' jk <- c(jk, join_keys(join_key("a", "b", "c")))
+#' jk <- c(jk, join_keys(join_key("a", "b", "c"), join_key("a", "b2", "c")))
+join_keys <- function(...) {
+  if (missing(...)) {
+    return(new_join_keys())
+  }
+  x <- rlang::list2(...)
+  if (length(x) > 1) {
+    return(join_keys.default(...))
+  }
+  UseMethod("join_keys", x[[1]])
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.join_keys <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.teal_data <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]@join_keys
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.TealData <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]$get_join_keys()
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.default <- function(...) {
+  x <- rlang::list2(...)
+  # Constructor
+  res <- new_join_keys()
+  if (length(x) > 0) {
+    join_keys(res) <- x
+  }
+
+  res
+}
+
+#' @rdname join_keys
+#' @details
+#' The setter assignment `join_keys(obj) <- ...` will merge obj and `...` if obj
+#' is not empty.
+#'
+#' @param join_keys_obj (`join_keys`) empty object to set the new relationship pairs.
+#' @param value (`join_key_set` or list of `join_key_set`) relationship pairs to add
+#' to `join_keys` list.
+#'
+#' @export
+`join_keys<-` <- function(join_keys_obj, value) {
+  UseMethod("join_keys<-", join_keys_obj)
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' # Using the setter (assignment) ----
+#'
+#' jk <- join_keys()
+#' join_keys(jk) <- join_key("ds1", "ds2", "some_col")
+#' join_keys(jk) <- c(join_keys(jk), join_key("ds3", "ds4", "some_col2"))
+#'
+#' join_keys(jk)[["ds1"]][["ds3"]] <- "some_col3"
+#' jk
+`join_keys<-.join_keys` <- function(join_keys_obj, value) {
+  if (missing(value)) {
```

Comment:
How does one call a replacement function with `value` missing? Is there an actual way other than `fun<-(<x>, <value>)`?

Reply:
No other way. it is a leftover from R6-based code that could be called without arguments

Done in 32a9b30c